_posts/2020-12-27-spring-boot-exception-hanlding.md

```diff
@@ -0,0 +1,596 @@
+According to its documentation, `ResponseEntityExceptionHandler` is a convenient base class for Controller Advice classes. It provides
+exception handlers for internal Spring exceptions. If we don't extend it then all the exceptions will be redirected to `DefaultHandlerExceptionResolver`
+which returns `ModelAndView` object. Since we are on the mission to handle all the exceptions
+by ourselves, we don't want to miss those.
+
+As you can see we have overridden two of the `ResponseEntityExceptionHandler` methods:
+* `handleMethodArgumentNotValid` - In [@ExceptionHandler](#exceptionhandler) section we had implemented a handler for it ourselves. In here we have only
+  overridden its behavior.
+* `handleExceptionInternal` - All the handlers in the `ResponseEntityExceptionHandler` use this function to build the
+  `ResponseEntity` similar to our `buildErrorResponse`. If we don't override this then the clients will receive only HTTP status
+  in response but since we include HTTP statues in our response bodies as well, we have overridden the method.
+
+
+<div class="notice warning">
+  <h4>Handling `NoHandlerFoundException` will requires few extra Steps</h4>
+  <p>
+  This exception occurs when you try to call the API which doesn't exist in the system. Despite us implementing it's handler 
+  via <code>ResponseEntityExceptionHandler</code> class the exception is redirected to <code>DefaultHandlerExceptionResolver</code>.
+  </p>
+  <p>
+  In order to redirect the exception to our adviser we need to set couple of properties in the the properties file: <code>spring.mvc.throw-exception-if-no-handler-found=true</code> and <code
+>spring.web.resources.add-mappings=false</code> 
+  </p>
+  <p>Credit: Stackoverflow user mengchengfeng <a href="https://stackoverflow.com/questions/36733254/spring-boot-rest-how-to-configure-404-resource-not-found">link</a>.</p>
+</div>
+
+### Some Points To Keep In Mind When Using @ControllerAdvice
+* To keep things simple always have only one Controller Adviser class in the project. It's good to have a single repository of
+  all the exceptions in the application. In case you create multiple then try to utilize `basePackages` or `annotations` property
+  to make it clear what controllers it's going to advise.
+* **Spring can process Controller Advice classes in any order** so be mindful when you write a catch-all handler in them. Especially
+  when you have not specified `basepPackages` or `annotations` in the annotation.
+
+## How Does Spring Process The Exceptions?
+Now that we have introduced ourselves with the mechanisms available to us for handling exceptions in Spring, let's
+understand in brief how Spring handles it and when one mechanism gets prioritized over the other.
+
+Please go through the following flow chart that traces the process of the exception handling by Spring:
+
+![Spring Exception Handling Flow](/assets/img/posts/spring-exception-handling/spring-exception-handling-mechanism.png)
```

Comment:
Nice! I love this chart! Can you make it look a bit more polished (i.e. without text overlapping the borders)? I know that's hard to do with the "diamond" form, but decisions don't necessarily need to be diamonds, they can be a (differently colored) box as well.